californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
The idea is to keep the previous behavior for this class with the new implementation.
Instead of getting notification through the `MessageObserver`, we get in now throw a notificationListener.
But we need to remove this notification listener when the observation is cancelled.